### DISCUSSION TITLE
How to remove the requirement that the workers need access to the DAG directory

### DISCUSSION THREAD

**Discussion by michalc** [apache/airflow#34650] [Q&A] [answered]
According to https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/overview.html it looks like workers need access to the DAG directory

![image](https://github.com/apache/airflow/assets/13877/fc2cfed7-0b77-43c8-a448-a246a838d6e3)

Is there a way to remove this requirement, say by writing a custom executor that parcels up the code the and sends it somewhere to be run? What would be involved in this? It is completely unfeasible?

It looks like this post from 2016 https://bytepawn.com/airflow.html suggests it was possible

> Airflow supports shipping the task’s code around to different nodes using pickle, 

but is this actually true?

**Comment by Taragolis** [comment, marked as answer]
> According to https://airflow.apache.org/docs/apache-airflow/stable/core-concepts/overview.html it looks like workers need access to the DAG directory

Yep, seems like the image a bit outdated, because webserver in Airflow 2.0 do not have need access to the DAG directory

> Is there a way to remove this requirement, say by writing a custom executor that parcels up the code the and sends it somewhere to be run? What would be involved in this? It is completely unfeasible?

In general executor send to worker that it should run [`airflow tasks run`](https://airflow.apache.org/docs/apache-airflow/stable/cli-and-env-variables-ref.html#run) cli command, so it would involve not only change on executor side but also in airflow core.

>> Airflow supports shipping the task’s code around to different nodes using pickle,

> but is this actually true?

Yes it is possible if executor support pickling, I guess only Celery and K8S support it. And I'm not sure about status of pickling in Airflow maybe it would remove in the future or maybe not

**Reply by michalc** [reply]
> Yes it is possible if executor support pickling, I guess only Celery and K8S support it.

Oh we’re already using the Celery executor! That certainly makes it simpler…

**Reply by michalc** [reply]
So I'm starting to suspect that the CeleryExecutor really does need access to the dags folder.

I've followed the instructions at https://airflow.apache.org/docs/apache-airflow/stable/howto/docker-compose/index.html, but I modified the worker definition in the docker-compose file so it doesn't have access to the dags folder

I then made a small dag in `my_dag.py` in the dags folder:

```python
from datetime import datetime
from airflow.decorators import dag, task

@dag(
    start_date=datetime(2022, 5, 16),
    catchup=False,
)
def MyDag():

    @task()
    def MyTask():
        print("HELLO")

    MyTask()

MyDag()
```
It appears in the web interface fine, But when I trigger it I see an error from the worker:

> airflow-test-airflow-worker-1     | [2023-09-28 14:51:21,866: WARNING/ForkPoolWorker-15] Dag 'MyDag' not found in path /opt/airflow/dags/my_dag.py; trying path /opt/airflow/dags

**Reply by Taragolis** [reply]
Just wondering do you set [`[core] donot_pickle`](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#donot-pickle) to `False`?

In additional I'm not sure how it work into the modern Airflow, there is probably couple of simple tests into the airflow core, and in general there is lack of interest to this feature from the community. So it might be buggy and not well documented.

**Reply by michalc** [reply]
> Just wondering do you set [[core] donot_pickle](https://airflow.apache.org/docs/apache-airflow/stable/configurations-ref.html#donot-pickle) to False?

I didn't before, but I have now (via the `AIRFLOW__CORE__DONOT_PICKLE` environment variable), and same thing.

I have also found the `--do-pickle` argument for the scheduler, and still same thing using it 

> In additional I'm not sure how it work into the modern Airflow, there is probably couple of simple tests into the airflow core, and in general there is lack of interest to this feature from the community. So it might be buggy and not well documented.

Ah understood

**Reply by Taragolis** [reply]
Just wondering, what is the problem with providing access to the DAG directory for the worker(s)?

**Reply by michalc** [reply]
Ah I'm on a platform/setup that has no shared disk between workers (or indeed, any processes), and where the update of DAG code is done by shutting down all workers, and starts them all up with the most recent code. This interrupts the workers, which isn't great on longer running tasks that can't pick up where they left off (at least, not without a lot of custom work per task).

There are a few ways to sort this, say some sort of syncing system from git where the DAG code lives that regularly updates the code on workers rather than re-deploying them, but it seems somehow neater/more robust for the scheduler to be the source of truth, and send off code for the workers to run. The workers don't need to deployed/interrupted on every code change and we can be sure that whatever the scheduler saw in terms of code is what the worker will be running.

**Reply by Taragolis** [reply]
Have you considered the opportunity about bundle DAGs into the Docker image?

If you ran on K8S cluster you might use [Airflow Helm Chart](https://airflow.apache.org/docs/helm-chart/stable/index.html) and it have [GitSync sidecar](https://airflow.apache.org/docs/helm-chart/stable/manage-dags-files.html#mounting-dags-using-git-sync-sidecar-with-persistence-enabled)

**Reply by michalc** [reply]
> Have you considered the opportunity about bundle DAGs into the Docker image?

So this isn't on Docker, but even if it were, then there is probably still the interruption problem - in most situations on deployment of a new image the old containers get shut down in a low number of minutes?

> [GitSync sidecar](https://airflow.apache.org/docs/helm-chart/stable/manage-dags-files.html#mounting-dags-using-git-sync-sidecar-with-persistence-enabled)

Suspect we will have to have some sort of git syncing process that's inspired by this.
